src/hdmf/container.py

```diff
@@ -753,12 +753,23 @@ def _generate_field_html(self, key, value, level, access_code):
     def _generate_array_html(self, array, level):
         """Generates HTML for array data"""
 
-        read_io = self.get_read_io()  # if the Container was read from file, get IO object
-        if read_io is not None: # Note that sometimes numpy array have a read_io attribute
-            repr_html = read_io.generate_dataset_html(array)
-        else:
+        is_numpy_array = isinstance(array, np.ndarray)
+        it_was_read_with_io = self.get_read_io() is not None
```

Comment:
`it_was_read_with_io` should capture most cases to determine if we need to route to the backend.  However, `it_was_read_with_io` applies to the whole container, but does not detect any changes the user may have done after reading.  As such, one case where this will potentially result in routing to the backend when we shouldn't is when the user replaces the dataset after read. However, I think adding a comment here to explain this should be sufficient.

Reply:
Ah, very good point about the user modifying the data on the fly.

The nature of the if-elif-else would catch that in most cases I think. Let's enumerate:
* If the user modified the data to be a numpy array it would be caught before in the first if.
* If the user modified the data to be DataIO it would be catch before in the second elif.
*If the user modified the data to be an Iterator then I think this will get here.

The Iterator is the tricky case, I was thinking already that maybe they should have a representation of their own. Do `AbstractDataChunkIterator` children have a shape and dtype attributes? If so, they will get here as these lines send things here for a representation.
```
 is_array_data = hasattr(value, "shape") and hasattr(value, "dtype")
```
(see the diff)

Reply:
> Do `AbstractDataChunkIterator` children have a shape and dtype attributes?

The iterators have a `dtype` but not a `shape`, because the final `shape` may not be known. However, they do have a `maxshape`. One option may be to add a html representation to `AbstractDataChunkIterator`

Reply:
Yes, I think adding a html represeentation to AbstractDataChunkIterator is something that I would like to do at some point if you are OK with it. Discussion for another PR though : )

Reply:
> Yes, I think adding a html represeentation to AbstractDataChunkIterator is something that I would like to do at some point if you are OK with it. Discussion for another PR though : )

Sure, that sounds good

Reply:
Looks good.